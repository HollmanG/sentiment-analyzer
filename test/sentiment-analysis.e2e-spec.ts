import { INestApplication, ValidationPipe } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import * as request from 'supertest';
import { AppModule } from '../src/app.module';

import mongoose from 'mongoose';

jest.setTimeout(60_000);

describe('SentimentAnalysisController (e2e)', () => {
  let app: INestApplication;

  beforeAll(async () => {
    const moduleRef: TestingModule = await Test.createTestingModule({
      imports: [AppModule],
    }).compile();

    await mongoose.connect(process.env.MONGO_URI);

    app = moduleRef.createNestApplication();

    app.useGlobalPipes(new ValidationPipe());

    await app.init();
  });

  afterAll(async () => {
    await mongoose.disconnect();
    await app.close();
  });

  beforeEach(async () => {
    await mongoose.connection.collection('SentimentAnalyses').deleteMany({});
  });

  describe('analyzeSentiment', () => {
    it('should return the sentiment analysis and save the result to the database', async () => {
      const sentence = 'I love programming!';

      const response = await request(app.getHttpServer())
        .post('/sentiment-analysis/analyze-sentiment')
        .send({ sentence })
        .expect(200);

      expect(response.body).toMatchSnapshot({
        sentence: expect.any(String),
        sentimentScore: expect.any(Number),
        sentimentMagnitude: expect.any(Number),
      });

      const savedAnalysis = await mongoose.connection.collection('SentimentAnalyses').findOne();

      expect(savedAnalysis).toMatchObject({
        sentence,
        sentimentScore: response.body.sentimentScore,
        sentimentMagnitude: response.body.sentimentMagnitude,
      });
    });

    it('should return a 400 error for invalid input', async () => {
      await request(app.getHttpServer()).post('/sentiment-analysis/analyze-sentiment').send({}).expect(400);
    });
  });

  describe('listAnalyses', () => {
    it('should return an array on analysis', async () => {
      await request(app.getHttpServer()).post('/sentiment-analysis/analyze-sentiment').send({ sentence: 'Yesterday was fun' });

      const analyses = await request(app.getHttpServer()).get('/sentiment-analysis/list-analyses').expect(200);

      analyses.body.forEach((analysis) => {
        expect(analysis).toMatchSnapshot({
          _id: expect.any(String),
          sentence: expect.any(String),
          createdAt: expect.any(String),
          sentimentScore: expect.any(Number),
          sentimentMagnitude: expect.any(Number),
        });
      });
    });
  });
});
