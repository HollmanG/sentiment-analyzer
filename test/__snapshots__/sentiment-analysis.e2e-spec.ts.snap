// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SentimentAnalysisController (e2e) analyzeSentiment should return the sentiment analysis and save the result to the database 1`] = `
{
  "sentence": Any<String>,
  "sentimentMagnitude": Any<Number>,
  "sentimentScore": Any<Number>,
}
`;

exports[`SentimentAnalysisController (e2e) listAnalyses should return an array on analysis 1`] = `
{
  "_id": Any<String>,
  "createdAt": Any<String>,
  "sentence": Any<String>,
  "sentimentMagnitude": Any<Number>,
  "sentimentScore": Any<Number>,
}
`;
